add_host: allow range of hosts (fix #41531)

rockandska opened the pull request:
##### SUMMARY

Add the ability to provide a range of hosts as "name/hosts" argument for add_host module. ( fix #41531 )

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME
```
lib/ansible/plugins/strategy/__init__.py
lib/ansible/plugins/action/add_host.py
lib/ansible/modules/inventory/add_host.py
```
##### ANSIBLE VERSION

```
ansible --version
ansible 2.7.0.dev0 (add_host_with_range 55eb86caa3) last updated 2018/06/16 15:37:35 (GMT +200)
  config file = None
  configured module search path = [u'/home/yoann/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/yoann/Scripts/ansible/lib/ansible
  executable location = /home/yoann/Scripts/ansible/bin/ansible
  python version = 2.7.12 (default, Nov 20 2017, 18:23:56) [GCC 5.4.0 20160609]
```


##### ADDITIONAL INFORMATION
Not sure if there will be side effect to always return "result" from add_host as a list, so i tried to stick to the actual way to return it if there is a single host to return.

Result with actual devel branch :
```
$ git checkout upstream/devel 

HEAD is now at 674dd3b... Add docstrings to meraki module utility methods

$ cat play_PR_41531 
---
- name: a
  hosts: localhost
  gather_facts: no
  tasks:
    - add_host:
        name: "{{ item }}"
      with_items:
        - 192.168.1.1
        - slave0:222
        - slave[0:1]
        - 192.168.1.[1:2]:24
    - debug:
        var: hostvars

$ ansible-playbook play_PR_41531 
 [WARNING]: Unable to parse /etc/ansible/hosts as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [a] **********************************************************************************************************************************************************************************************************

TASK [add_host] ***************************************************************************************************************************************************************************************************
changed: [localhost] => (item=192.168.1.1)
changed: [localhost] => (item=slave0:222)
changed: [localhost] => (item=slave[0:1])
changed: [localhost] => (item=192.168.1.[1:2]:24)

TASK [debug] ******************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "hostvars": {
        "192.168.1.1": {
            "ansible_check_mode": false, 
            "ansible_diff_mode": false, 
            "ansible_facts": {}, 
            "ansible_forks": 5, 
            "ansible_inventory_sources": [
                "/etc/ansible/hosts"
            ], 
            "ansible_playbook_python": "/usr/bin/python", 
            "ansible_run_tags": [
                "all"
            ], 
            "ansible_skip_tags": [], 
            "ansible_verbosity": 0, 
            "ansible_version": {
                "full": "2.7.0.dev0", 
                "major": 2, 
                "minor": 7, 
                "revision": 0, 
                "string": "2.7.0.dev0"
            }, 
            "group_names": [
                "ungrouped"
            ], 
            "groups": {
                "all": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave[0:1]", 
                    "192.168.1.[1:2]:24"
                ], 
                "ungrouped": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave[0:1]", 
                    "192.168.1.[1:2]:24"
                ]
            }, 
            "inventory_dir": null, 
            "inventory_file": null, 
            "inventory_hostname": "192.168.1.1", 
            "inventory_hostname_short": "192", 
            "omit": "__omit_place_holder__031f7ee4fb49d6906ca99b02893ec977d05d12b6", 
            "playbook_dir": "/home/yoann/Scripts/ansible"
        }, 
        "192.168.1.[1:2]:24": {
            "ansible_check_mode": false, 
            "ansible_diff_mode": false, 
            "ansible_facts": {}, 
            "ansible_forks": 5, 
            "ansible_inventory_sources": [
                "/etc/ansible/hosts"
            ], 
            "ansible_playbook_python": "/usr/bin/python", 
            "ansible_run_tags": [
                "all"
            ], 
            "ansible_skip_tags": [], 
            "ansible_verbosity": 0, 
            "ansible_version": {
                "full": "2.7.0.dev0", 
                "major": 2, 
                "minor": 7, 
                "revision": 0, 
                "string": "2.7.0.dev0"
            }, 
            "group_names": [
                "ungrouped"
            ], 
            "groups": {
                "all": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave[0:1]", 
                    "192.168.1.[1:2]:24"
                ], 
                "ungrouped": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave[0:1]", 
                    "192.168.1.[1:2]:24"
                ]
            }, 
            "inventory_dir": null, 
            "inventory_file": null, 
            "inventory_hostname": "192.168.1.[1:2]:24", 
            "inventory_hostname_short": "192", 
            "omit": "__omit_place_holder__031f7ee4fb49d6906ca99b02893ec977d05d12b6", 
            "playbook_dir": "/home/yoann/Scripts/ansible"
        }, 
        "slave0": {
            "ansible_check_mode": false, 
            "ansible_diff_mode": false, 
            "ansible_facts": {}, 
            "ansible_forks": 5, 
            "ansible_inventory_sources": [
                "/etc/ansible/hosts"
            ], 
            "ansible_playbook_python": "/usr/bin/python", 
            "ansible_run_tags": [
                "all"
            ], 
            "ansible_skip_tags": [], 
            "ansible_ssh_port": 222, 
            "ansible_verbosity": 0, 
            "ansible_version": {
                "full": "2.7.0.dev0", 
                "major": 2, 
                "minor": 7, 
                "revision": 0, 
                "string": "2.7.0.dev0"
            }, 
            "group_names": [
                "ungrouped"
            ], 
            "groups": {
                "all": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave[0:1]", 
                    "192.168.1.[1:2]:24"
                ], 
                "ungrouped": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave[0:1]", 
                    "192.168.1.[1:2]:24"
                ]
            }, 
            "inventory_dir": null, 
            "inventory_file": null, 
            "inventory_hostname": "slave0", 
            "inventory_hostname_short": "slave0", 
            "omit": "__omit_place_holder__031f7ee4fb49d6906ca99b02893ec977d05d12b6", 
            "playbook_dir": "/home/yoann/Scripts/ansible"
        }, 
        "slave[0:1]": {
            "ansible_check_mode": false, 
            "ansible_diff_mode": false, 
            "ansible_facts": {}, 
            "ansible_forks": 5, 
            "ansible_inventory_sources": [
            ], 
            "ansible_playbook_python": "/usr/bin/python", 
            "ansible_run_tags": [
                "all"
            ], 
            "ansible_skip_tags": [], 
            "ansible_verbosity": 0, 
            "ansible_version": {
                "full": "2.7.0.dev0", 
                "major": 2, 
                "minor": 7, 
                "revision": 0, 
                "string": "2.7.0.dev0"
            }, 
            "group_names": [
                "ungrouped"
            ], 
            "groups": {
                "all": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave[0:1]", 
                    "192.168.1.[1:2]:24"
                ], 
                "ungrouped": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave[0:1]", 
                    "192.168.1.[1:2]:24"
                ]
            }, 
            "inventory_dir": null, 
            "inventory_file": null, 
            "inventory_hostname": "slave[0:1]", 
            "inventory_hostname_short": "slave[0:1]", 
            "omit": "__omit_place_holder__031f7ee4fb49d6906ca99b02893ec977d05d12b6", 
            "playbook_dir": "/home/yoann/Scripts/ansible"
        }
    }
}

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0   
```

Result with this PR : 
```
$ git checkout add_host_with_range 
Previous HEAD position was 674dd3b... Add docstrings to meraki module utility methods
Switched to branch 'add_host_with_range'

$ git rev-parse HEAD
55eb86caa3fc42d8c1ec2daf00749843ad1c78f1

$ ansible-playbook -vv play_PR_41531 
ansible-playbook 2.7.0.dev0 (add_host_with_range 55eb86caa3) last updated 2018/06/16 15:37:35 (GMT +200)
  config file = None
  configured module search path = [u'/home/yoann/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/yoann/Scripts/ansible/lib/ansible
  executable location = /home/yoann/Scripts/ansible/bin/ansible-playbook
  python version = 2.7.12 (default, Nov 20 2017, 18:23:56) [GCC 5.4.0 20160609]
No config file found; using defaults
 [WARNING]: Unable to parse /etc/ansible/hosts as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAYBOOK: play_PR_41531 *******************************************************************************************************************************************************************************************
1 plays in play_PR_41531

PLAY [a] **********************************************************************************************************************************************************************************************************
META: ran handlers

TASK [add_host] ***************************************************************************************************************************************************************************************************
task path: /home/yoann/Scripts/ansible/play_PR_41531:6
creating host via 'add_host': hostname=192.168.1.1
changed: [localhost] => (item=192.168.1.1) => {"add_host": {"groups": [], "host_name": "192.168.1.1", "host_vars": {}}, "changed": true, "item": "192.168.1.1"}
creating host via 'add_host': hostname=slave0
changed: [localhost] => (item=slave0:222) => {"add_host": {"groups": [], "host_name": "slave0", "host_vars": {"ansible_ssh_port": 222}}, "changed": true, "item": "slave0:222"}
creating host via 'add_host': hostname=slave0
creating host via 'add_host': hostname=slave1
changed: [localhost] => (item=slave[0:1]) => {"add_host": [{"groups": [], "host_name": "slave0", "host_vars": {}}, {"groups": [], "host_name": "slave1", "host_vars": {}}], "changed": true, "item": "slave[0:1]"}
creating host via 'add_host': hostname=192.168.1.1
creating host via 'add_host': hostname=192.168.1.2
changed: [localhost] => (item=192.168.1.[1:2]:24) => {"add_host": [{"groups": [], "host_name": "192.168.1.1", "host_vars": {"ansible_ssh_port": 24}}, {"groups": [], "host_name": "192.168.1.2", "host_vars": {"ansible_ssh_port": 24}}], "changed": true, "item": "192.168.1.[1:2]:24"}

TASK [debug] ******************************************************************************************************************************************************************************************************
task path: /home/yoann/Scripts/ansible/play_PR_41531:13
ok: [localhost] => {
    "hostvars": {
        "192.168.1.1": {
            "ansible_check_mode": false, 
            "ansible_diff_mode": false, 
            "ansible_facts": {}, 
            "ansible_forks": 5, 
            "ansible_inventory_sources": [
                "/etc/ansible/hosts"
            ], 
            "ansible_playbook_python": "/usr/bin/python", 
            "ansible_run_tags": [
                "all"
            ], 
            "ansible_skip_tags": [], 
            "ansible_ssh_port": 24, 
            "ansible_verbosity": 2, 
            "ansible_version": {
                "full": "2.7.0.dev0", 
                "major": 2, 
                "minor": 7, 
                "revision": 0, 
                "string": "2.7.0.dev0"
            }, 
            "group_names": [
                "ungrouped"
            ], 
            "groups": {
                "all": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave1", 
                    "192.168.1.2"
                ], 
                "ungrouped": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave1", 
                    "192.168.1.2"
                ]
            }, 
            "inventory_dir": null, 
            "inventory_file": null, 
            "inventory_hostname": "192.168.1.1", 
            "inventory_hostname_short": "192", 
            "omit": "__omit_place_holder__4eb42ce5e76a7a5f68003ebe9c5d86e6bf0bb4f6", 
            "playbook_dir": "/home/yoann/Scripts/ansible"
        }, 
        "192.168.1.2": {
            "ansible_check_mode": false, 
            "ansible_diff_mode": false, 
            "ansible_facts": {}, 
            "ansible_forks": 5, 
            "ansible_inventory_sources": [
                "/etc/ansible/hosts"
            ], 
            "ansible_playbook_python": "/usr/bin/python", 
            "ansible_run_tags": [
                "all"
            ], 
            "ansible_skip_tags": [], 
            "ansible_ssh_port": 24, 
            "ansible_verbosity": 2, 
            "ansible_version": {
                "full": "2.7.0.dev0", 
                "major": 2, 
                "minor": 7, 
                "revision": 0, 
                "string": "2.7.0.dev0"
            }, 
            "group_names": [
                "ungrouped"
            ], 
            "groups": {
                "all": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave1", 
                    "192.168.1.2"
                ], 
                "ungrouped": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave1", 
                    "192.168.1.2"
                ]
            }, 
            "inventory_dir": null, 
            "inventory_file": null, 
            "inventory_hostname": "192.168.1.2", 
            "inventory_hostname_short": "192", 
            "omit": "__omit_place_holder__4eb42ce5e76a7a5f68003ebe9c5d86e6bf0bb4f6", 
            "playbook_dir": "/home/yoann/Scripts/ansible"
        }, 
        "slave0": {
            "ansible_check_mode": false, 
            "ansible_diff_mode": false, 
            "ansible_facts": {}, 
            "ansible_forks": 5, 
            "ansible_inventory_sources": [
                "/etc/ansible/hosts"
            ], 
            "ansible_playbook_python": "/usr/bin/python", 
            "ansible_run_tags": [
                "all"
            ], 
            "ansible_skip_tags": [], 
            "ansible_ssh_port": 222, 
            "ansible_verbosity": 2, 
            "ansible_version": {
                "full": "2.7.0.dev0", 
                "major": 2, 
                "minor": 7, 
                "revision": 0, 
                "string": "2.7.0.dev0"
            }, 
            "group_names": [
                "ungrouped"
            ], 
            "groups": {
                "all": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave1", 
                    "192.168.1.2"
                ], 
                "ungrouped": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave1", 
                    "192.168.1.2"
                ]
            "inventory_dir": null, 
            "inventory_file": null, 
            "inventory_hostname": "slave0", 
            "inventory_hostname_short": "slave0", 
            "omit": "__omit_place_holder__4eb42ce5e76a7a5f68003ebe9c5d86e6bf0bb4f6", 
            "playbook_dir": "/home/yoann/Scripts/ansible"
        }, 
        "slave1": {
            "ansible_check_mode": false, 
            "ansible_diff_mode": false, 
            "ansible_facts": {}, 
            "ansible_forks": 5, 
            "ansible_inventory_sources": [
                "/etc/ansible/hosts"
            ], 
            "ansible_playbook_python": "/usr/bin/python", 
            "ansible_run_tags": [
                "all"
            ], 
            "ansible_skip_tags": [], 
            "ansible_verbosity": 2, 
            "ansible_version": {
                "full": "2.7.0.dev0", 
                "major": 2, 
                "minor": 7, 
                "revision": 0, 
                "string": "2.7.0.dev0"
            }, 
            "group_names": [
                "ungrouped"
            ], 
            "groups": {
                "all": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave1", 
                    "192.168.1.2"
                ], 
                "ungrouped": [
                    "192.168.1.1", 
                    "slave0", 
                    "slave1", 
                    "192.168.1.2"
                ]
            }, 
            "inventory_dir": null, 
            "inventory_file": null, 
            "inventory_hostname": "slave1", 
            "inventory_hostname_short": "slave1", 
            "omit": "__omit_place_holder__4eb42ce5e76a7a5f68003ebe9c5d86e6bf0bb4f6", 
            "playbook_dir": "/home/yoann/Scripts/ansible"
        }
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0   
```

rockandska committed:
fix #41531 : allow result_item['add_host'] to be a list

Signed-off-by: RockAndSka <yoann_mac_donald@hotmail.com>

rockandska committed:
fix #41531 : return result_item['add_host'] as a list if range was provided

Signed-off-by: RockAndSka <yoann_mac_donald@hotmail.com>

rockandska committed:
fix #41531 : update add_host documentation

Signed-off-by: RockAndSka <yoann_mac_donald@hotmail.com>

rockandska committed:
Fix pep8 sanity check/1

Signed-off-by: RockAndSka <yoann_mac_donald@hotmail.com>

bcoca commented:
i don't see the need to push this into the `add_host` logic as you could just unroll the ranges in the loop

rockandska commented:
I already tried to achieve my need with loop and/or custom filter based on IRC discussions but wasn't able to reproduce the result i achieve with this PR with stable branch of ansible, that's why i thinked that this PR had an utility.

Even without my actual personal needs for this PR, i think it is kind of handy to be able to create range of host without adding any complexity with (custom) filters / loop.

My motivation for this PR  was triggered by the need to be able to transform this host_vars file ( i need to stick to this format even if it can sound weird) into the inventory bellow ( generated with this PR ).
If i miss something and someone have a working solution with stable branch to reproduce the exemples bellow, i will be really thankful to him/her.

Best regards,

**hostvars:**

```yaml
 /tmp/ansiblePR41678 $ cat host_vars/localhost 
---
hosts2configure:
  server1:
    hostname: server1
    groups:
      - linux
      - debian
    services:
      apache:
        state: present
      nginx:
        state: absent
  server2:
    hostname: server1
    groups:
      - linux
      - centos
    services:
      apache:
        state: absent
      nginx:
        state: present
  mysql[0:1]:222:
    groups:
      - ubuntu
      - linux
    services:
      mysql:
        state: absent
      mariadb:
        state: present
```

**play**
```yaml
 /tmp/ansiblePR41678 $ cat play.yml 
---
# Adding hosts
- hosts: localhost
  tasks:
    - add_host:
        name: "{{ item.key }}"
        created_by_host: "{{ inventory_hostname }}"
        created_by_key: "{{ item.key }}"
        groups: "{{ hostvars[inventory_hostname].hosts2configure[item.key]['groups'] | join(',') | default('all', true) }}"
      with_dict: "{{ hostvars[inventory_hostname].hosts2configure }}"

# Adding host_vars
- hosts: 00defaultgroup
  gather_facts: no
  tasks:
    - set_fact:
        "{{ item.key }}": "{{ item.value }}"
      with_dict: "{{ hostvars[created_by_host].hosts2configure[created_by_key] }}"

# use the new hosts
- hosts: all
  gather_facts: no
  tasks:
    - debug:
        var: hostvars[inventory_hostname]
```

**results**
```
 /tmp/ansiblePR41678  $  ansible-playbook -vvvv play.yml 
ansible-playbook 2.7.0.dev0 (add_host_with_range 5153959fee) last updated 2018/06/16 16:49:25 (GMT +200)
  config file = None
  configured module search path = [u'/home/yoann/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/yoann/Scripts/ansible/lib/ansible
  executable location = /home/yoann/Scripts/ansible/bin/ansible-playbook
  python version = 2.7.12 (default, Nov 20 2017, 18:23:56) [GCC 5.4.0 20160609]
No config file found; using defaults
setting up inventory plugins
 [WARNING]: Unable to parse /etc/ansible/hosts as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

Loading callback plugin default of type stdout, v2.0 from /home/yoann/Scripts/ansible/lib/ansible/plugins/callback/default.pyc

PLAYBOOK: play.yml **************************************************************************************************************************************************************************************************************************
3 plays in play.yml

PLAY [localhost] ****************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] **********************************************************************************************************************************************************************************************************************
task path: /tmp/ansiblePR41678/play.yml:2
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yoann
<127.0.0.1> EXEC /bin/sh -c 'echo ~yoann && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yoann/.ansible/tmp/ansible-tmp-1529345915.53-208128012335694 `" && echo ansible-tmp-1529345915.53-208128012335694="` echo /home/yoann/.ansible/tmp/ansible-tmp-1529345915.53-208128012335694 `" ) && sleep 0'
Using module file /home/yoann/Scripts/ansible/lib/ansible/modules/system/setup.py
<127.0.0.1> PUT /home/yoann/.ansible/tmp/ansible-local-25990tQIkDf/tmpLQAgNL TO /home/yoann/.ansible/tmp/ansible-tmp-1529345915.53-208128012335694/setup.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yoann/.ansible/tmp/ansible-tmp-1529345915.53-208128012335694/ /home/yoann/.ansible/tmp/ansible-tmp-1529345915.53-208128012335694/setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/yoann/.ansible/tmp/ansible-tmp-1529345915.53-208128012335694/setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yoann/.ansible/tmp/ansible-tmp-1529345915.53-208128012335694/ > /dev/null 2>&1 && sleep 0'
ok: [localhost]
META: ran handlers

TASK [add_host] *****************************************************************************************************************************************************************************************************************************
task path: /tmp/ansiblePR41678/play.yml:4
creating host via 'add_host': hostname=server1
changed: [localhost] => (item={'value': {u'services': {u'apache': {u'state': u'present'}, u'nginx': {u'state': u'absent'}}, u'hostname': u'server1', u'groups': [u'linux', u'debian']}, 'key': u'server1'}) => {
    "add_host": {
        "groups": [
            "linux", 
            "debian"
        ], 
        "host_name": "server1", 
        "host_vars": {
            "created_by_host": "localhost", 
            "created_by_key": "server1"
        }
    }, 
    "changed": true, 
    "item": {
        "key": "server1", 
        "value": {
            "groups": [
                "linux", 
                "debian"
            ], 
            "hostname": "server1", 
            "services": {
                "apache": {
                    "state": "present"
                }, 
                "nginx": {
                    "state": "absent"
                }
            }
        }
    }
}
creating host via 'add_host': hostname=server2
changed: [localhost] => (item={'value': {u'services': {u'apache': {u'state': u'absent'}, u'nginx': {u'state': u'present'}}, u'hostname': u'server1', u'groups': [u'linux', u'centos']}, 'key': u'server2'}) => {
    "add_host": {
        "groups": [
            "linux", 
            "centos"
        ], 
        "host_name": "server2", 
        "host_vars": {
            "created_by_host": "localhost", 
            "created_by_key": "server2"
        }
    }, 
    "changed": true, 
    "item": {
        "key": "server2", 
        "value": {
            "groups": [
                "linux", 
                "centos"
            ], 
            "hostname": "server1", 
            "services": {
                "apache": {
                    "state": "absent"
                }, 
                "nginx": {
                    "state": "present"
                }
            }
        }
    }
}
creating host via 'add_host': hostname=mysql0
creating host via 'add_host': hostname=mysql1
changed: [localhost] => (item={'value': {u'services': {u'mariadb': {u'state': u'present'}, u'mysql': {u'state': u'absent'}}, u'groups': [u'ubuntu', u'linux']}, 'key': u'mysql[0:1]:222'}) => {
    "add_host": [
        {
            "groups": [
                "ubuntu", 
                "linux"
            ], 
            "host_name": "mysql0", 
            "host_vars": {
                "ansible_ssh_port": 222, 
                "created_by_host": "localhost", 
                "created_by_key": "mysql[0:1]:222"
            }
        }, 
        {
            "groups": [
                "ubuntu", 
                "linux"
            ], 
            "host_name": "mysql1", 
            "host_vars": {
                "ansible_ssh_port": 222, 
                "created_by_host": "localhost", 
                "created_by_key": "mysql[0:1]:222"
            }
        }
    ], 
    "changed": true, 
    "item": {
        "key": "mysql[0:1]:222", 
        "value": {
            "groups": [
                "ubuntu", 
                "linux"
            ], 
            "services": {
                "mariadb": {
                    "state": "present"
                }, 
                "mysql": {
                    "state": "absent"
                }
            }
        }
    }
}
META: ran handlers
META: ran handlers
 [WARNING]: Could not match supplied host pattern, ignoring: 00defaultgroup


PLAY [00defaultgroup] ***********************************************************************************************************************************************************************************************************************
skipping: no hosts matched

PLAY [all] **********************************************************************************************************************************************************************************************************************************
META: ran handlers

TASK [debug] ********************************************************************************************************************************************************************************************************************************
task path: /tmp/ansiblePR41678/play.yml:23
ok: [server1] => {
    "hostvars[inventory_hostname]": {
        "ansible_check_mode": false, 
        "ansible_diff_mode": false, 
        "ansible_facts": {}, 
        "ansible_forks": 5, 
        "ansible_inventory_sources": [
            "/etc/ansible/hosts"
        ], 
        "ansible_playbook_python": "/usr/bin/python", 
        "ansible_run_tags": [
            "all"
        ], 
        "ansible_skip_tags": [], 
        "ansible_verbosity": 4, 
        "ansible_version": {
            "full": "2.7.0.dev0", 
            "major": 2, 
            "minor": 7, 
            "revision": 0, 
            "string": "2.7.0.dev0"
        }, 
        "created_by_host": "localhost", 
        "created_by_key": "server1", 
        "group_names": [
            "debian", 
            "linux"
        ], 
        "groups": {
            "all": [
                "server1", 
                "server2", 
                "mysql0", 
                "mysql1"
            ], 
            "centos": [
                "server2"
            ], 
            "debian": [
                "server1"
            ], 
            "linux": [
                "server1", 
                "server2", 
                "mysql0", 
                "mysql1"
            ], 
            "ubuntu": [
                "mysql0", 
                "mysql1"
            ], 
            "ungrouped": []
        }, 
        "inventory_dir": null, 
        "inventory_file": null, 
        "inventory_hostname": "server1", 
        "inventory_hostname_short": "server1", 
        "omit": "__omit_place_holder__8159371f1754d6452e9dda2ad2bff8c27ef84e37", 
        "playbook_dir": "/tmp/ansiblePR41678"
    }
}
ok: [server2] => {
    "hostvars[inventory_hostname]": {
        "ansible_check_mode": false, 
        "ansible_diff_mode": false, 
        "ansible_facts": {}, 
        "ansible_forks": 5, 
        "ansible_inventory_sources": [
            "/etc/ansible/hosts"
        ], 
        "ansible_playbook_python": "/usr/bin/python", 
        "ansible_run_tags": [
            "all"
        ], 
        "ansible_skip_tags": [], 
        "ansible_verbosity": 4, 
        "ansible_version": {
            "full": "2.7.0.dev0", 
            "major": 2, 
            "minor": 7, 
            "revision": 0, 
            "string": "2.7.0.dev0"
        }, 
        "created_by_host": "localhost", 
        "created_by_key": "server2", 
        "group_names": [
            "centos", 
            "linux"
        ], 
        "groups": {
            "all": [
                "server1", 
                "server2", 
                "mysql0", 
                "mysql1"
            ], 
            "centos": [
                "server2"
            ], 
            "debian": [
                "server1"
            ], 
            "linux": [
                "server1", 
                "server2", 
                "mysql0", 
                "mysql1"
            ], 
            "ubuntu": [
                "mysql0", 
                "mysql1"
            ], 
            "ungrouped": []
        }, 
        "inventory_dir": null, 
        "inventory_file": null, 
        "inventory_hostname": "server2", 
        "inventory_hostname_short": "server2", 
        "omit": "__omit_place_holder__8159371f1754d6452e9dda2ad2bff8c27ef84e37", 
        "playbook_dir": "/tmp/ansiblePR41678"
    }
}
ok: [mysql0] => {
    "hostvars[inventory_hostname]": {
        "ansible_check_mode": false, 
        "ansible_diff_mode": false, 
        "ansible_facts": {}, 
        "ansible_forks": 5, 
        "ansible_inventory_sources": [
            "/etc/ansible/hosts"
        ], 
        "ansible_playbook_python": "/usr/bin/python", 
        "ansible_run_tags": [
            "all"
        ], 
        "ansible_skip_tags": [], 
        "ansible_ssh_port": 222, 
        "ansible_verbosity": 4, 
        "ansible_version": {
            "full": "2.7.0.dev0", 
            "major": 2, 
            "minor": 7, 
            "revision": 0, 
            "string": "2.7.0.dev0"
        }, 
        "created_by_host": "localhost", 
        "created_by_key": "mysql[0:1]:222", 
        "group_names": [
            "linux", 
            "ubuntu"
        ], 
        "groups": {
            "all": [
                "server1", 
                "server2", 
                "mysql0", 
                "mysql1"
            ], 
            "centos": [
                "server2"
            ], 
            "debian": [
                "server1"
            ], 
            "linux": [
                "server1", 
                "server2", 
                "mysql0", 
                "mysql1"
            ], 
            "ubuntu": [
                "mysql0", 
                "mysql1"
            ], 
            "ungrouped": []
        }, 
        "inventory_dir": null, 
        "inventory_file": null, 
        "inventory_hostname": "mysql0", 
        "inventory_hostname_short": "mysql0", 
        "omit": "__omit_place_holder__8159371f1754d6452e9dda2ad2bff8c27ef84e37", 
        "playbook_dir": "/tmp/ansiblePR41678"
    }
}
ok: [mysql1] => {
    "hostvars[inventory_hostname]": {
        "ansible_check_mode": false, 
        "ansible_diff_mode": false, 
        "ansible_facts": {}, 
        "ansible_forks": 5, 
        "ansible_inventory_sources": [
            "/etc/ansible/hosts"
        ], 
        "ansible_playbook_python": "/usr/bin/python", 
        "ansible_run_tags": [
            "all"
        ], 
        "ansible_skip_tags": [], 
        "ansible_ssh_port": 222, 
        "ansible_verbosity": 4, 
        "ansible_version": {
            "full": "2.7.0.dev0", 
            "major": 2, 
            "minor": 7, 
            "revision": 0, 
            "string": "2.7.0.dev0"
        }, 
        "created_by_host": "localhost", 
        "created_by_key": "mysql[0:1]:222", 
        "group_names": [
            "linux", 
            "ubuntu"
        ], 
        "groups": {
            "all": [
                "server1", 
                "server2", 
                "mysql0", 
                "mysql1"
            ], 
            "centos": [
                "server2"
            ], 
            "debian": [
                "server1"
            ], 
            "linux": [
                "server1", 
                "server2", 
                "mysql0", 
                "mysql1"
            ], 
            "ubuntu": [
                "mysql0", 
                "mysql1"
            ], 
            "ungrouped": []
        }, 
        "inventory_dir": null, 
        "inventory_file": null, 
        "inventory_hostname": "mysql1", 
        "inventory_hostname_short": "mysql1", 
        "omit": "__omit_place_holder__8159371f1754d6452e9dda2ad2bff8c27ef84e37", 
        "playbook_dir": "/tmp/ansiblePR41678"
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP **********************************************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0   
mysql0                     : ok=1    changed=0    unreachable=0    failed=0   
mysql1                     : ok=1    changed=0    unreachable=0    failed=0   
server1                    : ok=1    changed=0    unreachable=0    failed=0   
server2                    : ok=1    changed=0    unreachable=0    failed=0   
```

samyscoub commented:
:+1:

bcoca commented:
An example of the op w/o needing code changes:
```
    - add_host:
         name: "{{ item }}"
      with_items:
      - 192.168.1.1
      - "{{range(0, 222)|map('regex_replace', '^(.*)$', 'slave\\1')| list}}"
      - "{{range(0, 1)|map('regex_replace', '^(.*)$', 'slave\\1')| list}}"
      - "{{range(1, 2)|map('regex_replace', '^(.*)$', '192.168.1.\\1:24')| list}}"
```

rockandska commented:
Closing after having found a solution for my use case by writing a small custom filter as suggest by @bcoca / @agaffney on IRC to preserve the original "key" somewhere and be able to access the attributes corresponding to my key later

```
from ansible.parsing.utils.addresses import parse_address
from ansible.plugins.inventory import detect_range, expand_hostname_range

class FilterModule(object):
    def filters(self):
        return {
            'hostslist2dict': self.hostslist2dict,
        }
 
    def hostslist2dict(self, hostslist):
        result_dict = {}
        for host in hostslist:
            hostnames = []
            name, port = parse_address(host, allow_ranges=True)
            if detect_range(name):
                hostnames = expand_hostname_range(name)
            else:
                hostnames = [name]
            for name in hostnames:
                if port:
                    name = name + ":" + str(port)
                result_dict.update( { name : { 'created_by' : host } } )
        return result_dict
```

```
---
# Adding hosts
- hosts: localhost
  tasks:
    - add_host:
        name: "{{ item.key }}"
        created_by_host: "{{ inventory_hostname }}"
        created_by_key: "{{ item.value.created_by }}"
        groups: "00defaultgroup, {{ hosts2monitor[item.value.created_by]['groups'] | default(['all'], True) | join(',') | lower }}"
      with_dict: "{{ hosts2configure.keys() | hostslist2dict }}"
```

Regards,
